aten/src/ATen/native/sparse/SparseFactories.h

```diff
@@ -0,0 +1,118 @@
+#pragma once
+#include <ATen/TensorIndexing.h>
+#include <ATen/TensorIterator.h>
+#include <ATen/core/ATen_fwd.h>
+#include <ATen/core/Tensor.h>
+#include <ATen/native/TensorFactories.h>
+#include <c10/core/Scalar.h>
+#include <c10/util/ArrayRef.h>
+#include <c10/util/Exception.h>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#include <ATen/NativeFunctions.h>
+#else
+#include <ATen/ops/sparse_coo_tensor.h>
+#endif
+
+namespace at {
+namespace native {
+namespace impl {
+
+template <typename setup_func_t, typename kernel_func_t>
+Tensor spdiags_impl(
+    const Tensor& diagonals,
+    const Tensor& offsets,
+    IntArrayRef shape,
+    c10::optional<Layout> layout,
+    const setup_func_t& setup_func,
+    const kernel_func_t& kernel_func) {
+  auto diagonals_2d = diagonals.dim() == 1 ? diagonals.unsqueeze(0) : diagonals;
+  TORCH_CHECK(diagonals_2d.dim() == 2, "Diagonals must be vector or matrix");
+  TORCH_CHECK(shape.size() == 2, "Output shape must be 2d");
+  auto offsets_1d = offsets.dim() == 0 ? offsets.unsqueeze(0) : offsets;
+  TORCH_CHECK(offsets_1d.dim() == 1, "Offsets must be scalar or vector");
+  TORCH_CHECK(
+      diagonals_2d.size(0) == offsets_1d.size(0),
+      "Number of diagonals (",
+      diagonals_2d.size(0),
+      ") does not match the number of offsets (",
+      offsets_1d.size(0),
+      ")");
+  if (layout) {
+    TORCH_CHECK(
+        (*layout == Layout::Sparse) || (*layout == Layout::SparseCsc) ||
+            (*layout == Layout::SparseCsr),
+        "Only output layouts (",
+        Layout::Sparse,
+        ", ",
+        Layout::SparseCsc,
+        ", and ",
+        Layout::SparseCsr,
+        ") are supported");
+  }
+  TORCH_CHECK(
+      offsets_1d.scalar_type() == at::kLong,
+      "spdiags(): Expected a LongTensor of offsets but got ",
+      offsets_1d.scalar_type());
+
+  auto nnz_per_diag = at::zeros_like(offsets_1d);
+
+  auto setup_iter = TensorIteratorConfig()
+                        .add_output(nnz_per_diag)
+                        .add_input(offsets_1d)
+                        .build();
+  int64_t nnz = 0;
+  // Checks offsets and computes nnz per diag
+  setup_func(setup_iter, shape[0], shape[1], diagonals_2d.size(1), nnz);
+  auto n_diag = offsets_1d.size(0);
+  auto nnz_prefix = nnz_per_diag.cumsum(0).sub(nnz_per_diag);
+
+  // We need to restride outputs so that the nnz dimension appears to be n_diag,
+  // and we will advance it manually
+  auto indices = at::zeros({2, nnz}, offsets_1d.options());
+  auto row_indices_restrided = indices[0].as_strided({n_diag}, {0});
+  auto col_indices_restrided = indices[1].as_strided({n_diag}, {0});
```

Comment:
Why? There is no need in supplying `indices` into a TI. It is sufficient that a TI has a single tensor of length `n_diag`, then the `loop` lambda captures `indices[0].data_ptr<int64_t>()` and `indices[1].data_ptr<int64_t>()`. Same goes for values.

Reply:
The output tensors (`values` and `indices`) are `(nnz,)` and `(2, nnz)`. So I had to provide a dummy output to avoid the explicit restride,  as `cpu_kernel` requires that the `iter` have one output.

Reply:
You can slice indices into rows and cols so that they are of the right size.

Reply:
Yes, there is this limitation. We need to check whether the dummy output is not blown out to mach the size of the inputs...